### ISSUE TITLE
Pyright incorrectly infers callable properties as methods

### ISSUE THREAD

**Issue by WSH032** [microsoft/pyright#7018]
**Describe the bug**
Pyright incorrectly infers `callable properties accepting positional arguments` as `methods`, 
but `callable properties accepting only keyword arguments` are inferred correctly.

**Code or Screenshots**
```py
from typing import Any, Callable
from typing_extensions import reveal_type


def foo(a: int, *, b: int):
    print(a, b)

def bar(*, a: int, b: int):
    print(a, b)


class A:
    foo: Callable[..., Any] = foo

class B:
    foo = foo
    bar = bar


# no complaints
(A.foo(1, b=2))
reveal_type(A.foo)  # (variable) def foo(...) -> Any

# error: Cannot access member "foo" for type "type[B]"
# Could not bind method "foo" because "type[B]" is not assignable to parameter "a"
#     "B" is incompatible with "int" (reportGeneralTypeIssues)
B.foo(1, b=2)
reveal_type(B.foo)  # Unknown

# no complaints
B.bar(a=1, b=2)
reveal_type(B.bar)  # (*, a: int, b: int) -> None
```

**VS Code extension or command-line**
 The command-line tool `v1.1.36`

**Comment by LeeeeT** [comment]
```python
class B:
    foo = foo
    bar = bar
```
is effectively the same as
```python
class B:
    def foo(a: int, *, b: int):
        print(a, b)

    def bar(*, a: int, b: int):
        print(a, b)
```
at runtime.

`foo` and `bar` become bound methods when accessed from an instance of `B`. That is because functions in python are actually descriptors doing the binding.

You can run pyright for the latter code snippet and see an error for `foo` and no errors for `bar`, just as in your original comment.

Pyright knows that functions are descriptors, and when they are members of a class, pyright expects them to take a `self` object as the first argument.

The reason why there's no error if you add `Callable[..., Any]` annotation is because `Callable` is a broader type than function type, and `Callable` by itself is **not** a descriptor. By writing `Callable[..., Any]` annotation you force pyright to "forget" that `foo` is a descriptor. Hence, no `self` is passed to it (in type checking world).

Type checking descriptors stuff is very tricky, and it is decided to just ignore it.

---

Why pyright allows
```python
class B:
    def bar(*, a: int, b: int):
        print(a, b)
```
makes me wonder tho.

**Comment by erictraut** [comment]
Pyright is working as designed here. As @LeeeeT noted, functions within a class are actually descriptors. The symbols `B.foo` and `B.bar` are effectively instance methods. If you attempt to call them through an instance of `B`, the first parameter will receive a "self" value.

```python
def foo(a: int, *, b: int):
    print(a, b)

class B:
    foo = foo

B().foo(b=2) # Prints "<B object> 2"
```

You might wonder why pyright attempts to bind `B` to `foo` when accessing `foo` through the class rather than an instance. That's important in cases where the class is generic. Consider the following example:

```python
T = TypeVar("T")

class C(Generic[T]):
    def method(self, x: T) -> T:
        return x

bound_method = C[int].method
reveal_type(bound_method) # (self: C[int], x: int) -> int
bound_method(C(), "") # Type error
```

In your example, when pyright attempts to bind `B` to `foo`, it is unable to do so without a type violation because `B` is not compatible with `int`.

If your intent is for `foo` and `bar` to be static methods in the `B` class, you can wrap `foo` and `bar` with a `staticmethod` call, like this:

```python
class B:
    foo = staticmethod(foo)
    bar = staticmethod(bar)
```

**Comment by WSH032** [comment]
Great! Thank you both for excellent explanations!

To be honest, I did overlook the case of `B().foo`.

---

> You might wonder why pyright attempts to bind B to foo when accessing foo through the class rather than an instance. That's important in cases where the class is generic. Consider the following example:

When calling `B.foo()`, pyright defaults to binding `self` to `foo`, that really confuses me before hearing your explanation.

However, I can still accept this point, because `staticmethod` is indeed more Pythonic.
